hpx/lcos/channel.hpp

```diff
@@ -0,0 +1,397 @@
+    {
+        typedef hpx::util::iterator_facade<
+                channel_iterator<void, Channel>, util::unused_type const,
+                std::input_iterator_tag
+            > base_type;
+
+    public:
+        channel_iterator()
+          : channel_(nullptr), data_(false)
+        {}
+
+        channel_iterator(Channel const& c)
+          : channel_(&c), data_(get_checked())
+        {}
+
+    private:
+        bool get_checked()
+        {
+            hpx::future<void> f = channel_->get_async();
+            f.wait();
+            return !f.has_exception();
+        }
+
+        friend class hpx::util::iterator_core_access;
+
+        bool equal(channel_iterator const& rhs) const
+        {
+            return (data_ == rhs.data_ &&
+                        (channel_ == rhs.channel_ ||
+                            (channel_ != nullptr && rhs.channel_ != nullptr &&
+                             channel_->get_id() == rhs.channel_->get_id())
+                        )
+                    ) ||
+                (!data_ && rhs.channel_ == nullptr) ||
+                (channel_ == nullptr && !rhs.data_);
+        }
+
+        void increment()
+        {
+            if (channel_)
+                data_ = get_checked();
+        }
+
+        typename base_type::reference dereference() const
+        {
+            HPX_ASSERT(data_);
+            return util::unused;
+        }
+
+    private:
+        Channel const* channel_;
+        bool data_;
+    };
+
+    ///////////////////////////////////////////////////////////////////////////
+    template <typename T>
+    class channel
+      : public components::client_base<channel<T>, lcos::server::channel<T> >
+    {
+        typedef components::client_base<
+                channel<T>, lcos::server::channel<T>
+            > base_type;
+
+        HPX_CONSTEXPR_OR_CONST
+        static std::size_t default_generation = std::size_t(-1);
+
+    public:
+        channel()
+        {}
+
+        // create a new instance of a channel component
+        channel(naming::id_type const& loc)
+          : base_type(hpx::new_<lcos::server::channel<T> >(loc))
+        {}
+        channel(lcos::future<naming::id_type> f)
+          : base_type(std::move(f))
+        {}
+
+        ///////////////////////////////////////////////////////////////////////
+        hpx::future<T>
+        get_async(std::size_t generation = default_generation) const
+        {
+            typedef typename lcos::server::channel<T>::get_generation_action
+                action_type;
+            return hpx::async(action_type(), this->get_id(), generation);
+        }
+        T get(std::size_t generation = default_generation,
+            hpx::error_code& ec = hpx::throws) const
+        {
+            return get_async(generation).get(ec);
+        }
+        T get(hpx::error_code& ec,
+            std::size_t generation = default_generation) const
+        {
+            return get_async(generation).get(ec);
+        }
```

Comment:
I was thinking of re-using the execution policies for this as well, it's a compelling thought. It would be a lot of work, though, as we have a lot of spots where this could be applied...